keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -0,0 +1,1354 @@
+  uid: foobar-uid
+...
+spec:
+  minimumRuntimeVersion: v12.0.0
+  minimumMemory: 32Gi
+  # "sharing" is a configuration parameter that does not
+  # get translated into the selector below.
+  sharing:
+    strategy: TimeSliced
+```
+
+Note that all fields in this CRD can be fully validated since it is owned by
+the DRA driver itself. This includes value ranges that are specific to the
+underlying hardware. There's no risk of using invalid attribute names because
+only the fields shown here are valid.
+
+With this CRD in place, a DRA driver controller is able to convert instances of
+it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
+is able to understand.
+
+For the example above, the converted object would look as follows:
+
+```yaml
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  # A vendor can put any kind of object here to pass the configuration
+  # parameters down to the kubelet plugin. In this case, the vendor
+  # driver simply copied the entire CR. It could also be some
+  # separate, smaller configuration type.
+  #
+  # Beware that ResourceClaimParameters have separate RBAC rules than
+  # the vendor CRD, so information included here may get visible
+  # to more users than the original CRD. Both objects are in the same
+  # namespace.
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 32Gi
+    sharing:
+      strategy: TimeSliced
+
+requests:
+- driverName: cards.dra.example.com
+  namedResourcesWithAttributes:
+    required:
+    # Selectors are CEL expressions with access to the attributes of the named resource
+    # that is being checked for a match. Each entry here is a request for one resource.
+    - selector: |-
+        attributes["runtimeVersion"] >= "v12.0.0" && attributes["memory"] >= "32Gi"
+```
+
+The meaning is that the selector expression must evaluate to true for a
+particular named resource in `namedResourcesWithAttributes`.
+
+Future extensions could be added to support partioning of resources as well as a
+express constraints that must be satisfied *between* any selected resources. For
+example, selecting two cards which are on the same PCI root complex may be
+needed to get the required performance.
+
+Instead of defining a vendor-specific CRD, DRA driver authors (or
+administrators) could decide to allow users to create and reference
+`ResourceClaimParameters` directly within their `ResourceClaims`. This would
```

Comment:
How does the ResourceClaim look like?
How does it reference the claim?

Reply:
The claim would reference the parameter object with "resource.k8s.io" as APIGroup, "ResourceClaimParameters" as kind, and then the name of the object. This object then gets used directly.